spring-integration-core/src/main/java/org/springframework/integration/channel/ExecutorChannel.java

```diff
@@ -124,15 +106,33 @@ public final void onInit() {
 					new BeanFactoryChannelResolver(this.getBeanFactory()));
 			this.executor = new ErrorHandlingTaskExecutor(this.executor, errorHandler);
 		}
-		this.dispatcher = new UnicastingDispatcher(this.executor);
-		this.dispatcher.setFailover(this.failover);
+		UnicastingDispatcher unicastingDispatcher = new UnicastingDispatcher(this.executor);
+		unicastingDispatcher.setFailover(this.failover);
 		if (this.maxSubscribers == null) {
-			this.maxSubscribers = this.getIntegrationProperty(IntegrationProperties.CHANNELS_MAX_UNICAST_SUBSCRIBERS, Integer.class);
+			this.maxSubscribers =
+					getIntegrationProperty(IntegrationProperties.CHANNELS_MAX_UNICAST_SUBSCRIBERS, Integer.class);
 		}
-		this.dispatcher.setMaxSubscribers(this.maxSubscribers);
+		unicastingDispatcher.setMaxSubscribers(this.maxSubscribers);
 		if (this.loadBalancingStrategy != null) {
-			this.dispatcher.setLoadBalancingStrategy(this.loadBalancingStrategy);
+			unicastingDispatcher.setLoadBalancingStrategy(this.loadBalancingStrategy);
 		}
+
+		unicastingDispatcher.setMessageHandlingTaskDecorator(new MessageHandlingTaskDecorator() {
+
+			@Override
+			public Runnable decorate(MessageHandlingRunnable task) {
+				if (!ExecutorChannel.this.executorInterceptors.isEmpty()) {
```

Comment:
See the optimization I did for INT-3719; calling `size()` was much more expensive than looking up a field. However, that might have been because it was a `CopyOnWriteArrayList`.

However, yours is a simple `ArrayList` - should it be a COWAL?

Not saying it's wrong but we need to be very sensitive about putting any unnecessary code in the main message flow.